app/lib/session_encryptor.rb

```diff
@@ -0,0 +1,59 @@
+class SessionEncryptor
+  KEY_LEN = ActiveSupport::MessageEncryptor.key_len
+
+  def initialize(session_signing_key:)
+    @session_signing_key = session_signing_key
+  end
+
+  def encrypt_session(access_token:, refresh_token:)
+    encrypt_string(
+      {
+        access_token: Base64.encode64(access_token),
+        refresh_token: Base64.encode64(refresh_token),
+      }.to_json,
+    )
+  end
+
+  def decrypt_session(ciphertext)
+    return if ciphertext.blank?
+
+    plaintext = decrypt_string(ciphertext)
+    return unless plaintext
+
+    hash = JSON.parse(plaintext)
+
+    {
+      access_token: Base64.decode64(hash["access_token"]),
+      refresh_token: Base64.decode64(hash["refresh_token"]),
+    }
+  end
+
+protected
+
+  attr_reader :session_signing_key
+
+  def encrypt_string(plaintext)
+    salt = SecureRandom.hex KEY_LEN
+    ciphertext = encryptor(salt).encrypt_and_sign(plaintext)
+    Base64.urlsafe_encode64("#{salt}$$#{ciphertext}")
+  end
+
+  def decrypt_string(ciphertext)
+    bits = Base64.urlsafe_decode64(ciphertext).split("$$")
+    return nil unless bits.length == 2
```

Comment:
I assume we expect only 2 bits because we know that `"$$"` is never present in `salt` nor in `ciphertext`. I get that for `salt` because it's a random hexadecimal string. Is that the case for `ciphertext` too?

Reply:
Yes, `ciphertext` is base64 encoded: https://apidock.com/rails/ActiveSupport/MessageEncryptor

Reply:
Ah, [I see](https://en.wikipedia.org/wiki/Base64#Base64_table)!